pkg/networkservice/common/connect/server.go

```diff
@@ -65,22 +74,44 @@ func NewServer(
 }
 
 func (s *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	conn, err := s.client(ctx, request.GetConnection()).Request(ctx, request.Clone())
+	clientURL := clienturlctx.ClientURL(ctx)
+	if clientURL == nil {
+		return nil, errors.Errorf("clientURL not found for incoming connection: %+v", request.GetConnection())
+	}
+
+	clientInfo := s.client(ctx, request.GetConnection())
+	conn, err := clientInfo.client.Request(ctx, request.Clone())
 	if err != nil {
+		if _, ok := s.connInfos.Load(request.GetConnection().GetId()); !ok {
+			s.closeClient(clientInfo, clientURL.String())
+		}
+
+		// Close current client chain if grpc connection was closed
+		if grpcutils.UnwrapCode(err) == codes.Canceled {
+			s.deleteClient(clientInfo, clientURL.String())
+			s.connInfos.Delete(request.GetConnection().GetId())
+		}
```

Comment:
Just because we get a codes.Canceled ... doesn't mean anything about the *client* ... it only tells you about the *connection*

Reply:
@edwarnicke That is the point, we want to close client if something wrong with connection. Right now if grpc connection is broken, we just fail all requests comming to this client and doing nothing to reconnect.

Reply:
Doesn't codes.Canceled mean something is wrong with the *request* not the *connection*?  I presume it comes back if the ctx for the request exceeds its timeout for example...

Reply:
I still suspect this will unnecessarily cycle connections when we don't need to... However, the blast radius should be fine.  Let's move on :)